tasks/eva_3667/src/main/groovy/eva3667/remediate_lowercase_nucleotide.groovy

```diff
@@ -0,0 +1,362 @@
+import uk.ac.ebi.eva.commons.models.mongo.entity.subdocuments.VariantStatsMongo
+import uk.ac.ebi.eva.commons.mongodb.entities.subdocuments.HgvsMongo
+
+import java.nio.file.Paths
+import java.util.stream.Collectors
+
+import static org.springframework.data.mongodb.core.query.Criteria.where
+
+def cli = new CliBuilder()
+cli.workingDir(args: 1, "Path to the working directory where processing files will be kept", required: true)
+cli.envPropertiesFile(args: 1, "Properties file with db details to use for remediation", required: true)
+cli.dbName(args: 1, "Database name that needs to be remediated", required: true)
+def options = cli.parse(args)
+if (!options) {
+    cli.usage()
+    System.exit(1)
+}
+
+
+// Run the remediation application
+new SpringApplicationBuilder(RemediationApplication.class).properties([
+        'spring.config.location'      : options.envPropertiesFile,
+        'spring.data.mongodb.database': options.dbName])
+        .run(options.workingDir, options.dbName)
+
+
+@SpringBootApplication
+class RemediationApplication implements CommandLineRunner {
+    private static Logger logger = LoggerFactory.getLogger(RemediationApplication.class)
+
+    public static final String VARIANTS_COLLECTION = "variants_2_0"
+    public static final String FILES_COLLECTION = "files_2_0"
+    public static final String ANNOTATIONS_COLLECTION = "annotations_2_0"
+    public static final String FILES_COLLECTION_STUDY_ID_KEY = "sid"
+    public static final String FILES_COLLECTION_FILE_ID_KEY = "fid"
+
+    public static final String REGEX_PATTERN = '(?:\\S+_){2}([ACGTN]*[^ACGTN_]+[ACGTN]*)?_([ACGTN]*[^ACGTN_]+[ACGTN]*)?(?=\\s*$)'
+
+    @Autowired
+    MongoTemplate mongoTemplate
+
+    @Override
+    void run(String... args) throws Exception {
+        String workingDir = args[0]
+        String dbName = args[1]
+
+        // workaround to not saving a field by the name _class (contains the java class name) in the document
+        MappingMongoConverter converter = mongoTemplate.getConverter()
+        converter.setTypeMapper(new DefaultMongoTypeMapper(null))
+
+        MongoCollection<VariantDocument> variantsColl = mongoTemplate.getCollection(VARIANTS_COLLECTION)
+        // Create a query to find all matching lowercase variants
+        Query regexQuery = new Query(where("_id").regex(REGEX_PATTERN))
+        // Obtain a MongoCursor to iterate through documents
+        def mongoCursor = variantsColl.find(regexQuery.getQueryObject()).iterator()
+
+        // Iterate through each variant one by one
+        while (mongoCursor.hasNext()) {
+            // read the variant as a VariantDocument
+            VariantDocument lowercaseVariant = mongoTemplate.getConverter().read(VariantDocument.class, mongoCursor.next())
+            logger.info("Processing Variant: {}", lowercaseVariant)
+
+            // filter out false positives - double check if ref or alt really contains lowercase
+            if (hasLowerCaseRefOrAlt(lowercaseVariant)) {
```

Comment:
It's really a style preference but I prefer not to have so many if/else indented in each other but to test each case separately and use `continue`
```java
            if (! hasLowerCaseRefOrAlt(lowercaseVariant)) {
                // False positive - variant captured by regex but does not have any lowercase character in ref or alt
                logger.info("Variant does not contain any lowercase ref or alt. variant_id: {}, variant_ref: {}, " +
                        "variant_alt: {}", lowercaseVariant.getId(), lowercaseVariant.getReference(),
                        lowercaseVariant.getAlternate()
                )
                continue
            }
            ...
            if (variantInDB == null) {
                 logger.warn("Variant with id {} not found in DB", newId)
                 remediateCaseNoIdCollision(lowercaseVariant, newId)
                 continue
             }
```

Reply:
done